src/packages/cli/src/detach.ts

```diff
@@ -0,0 +1,261 @@
+    startTime,
+    pid,
+    instanceName: createInstanceName(),
+    host,
+    port,
+    flavor,
+    cmd,
+    version
+  };
+
+  while (true) {
+    const instanceFilename = await getInstanceFilePath(instance.instanceName);
+    try {
+      await writeFile(instanceFilename, JSON.stringify(instance), {
+        // wx means "Open file for writing, but fail if the path exists". see
+        // https://nodejs.org/api/fs.html#file-system-flags
+        flag: "wx"
+      });
+      break;
+    } catch (err) {
+      switch ((err as CodedError).code) {
+        case "EEXIST":
+          // an instance already exists with this name
+          instance.instanceName = createInstanceName();
+          break;
+        case "ENOENT":
+          // we don't check whether the folder exists before writing, as that's
+          // a very uncommon case. Catching the exception and subsequently
+          // creating the directory is faster in the majority of cases.
+          await mkdir(dataPath, { recursive: true });
+          break;
+        default:
+          throw err;
+      }
+    }
+  }
+
+  return instance;
+}
+
+/**
+ * Fetch all instance of Ganache running in detached mode. Cleans up any
+ * instance files for processes that are no longer running.
+ * @returns {Promise<DetachedInstance[]>} resolves with an array of instances
+ */
+export async function getDetachedInstances(): Promise<DetachedInstance[]> {
+  let files: string[];
+  let processes: ProcessDescriptor[];
+  try {
+    [files, processes] = await Promise.all([readdir(await dataPath), psList()]);
+  } catch (err) {
+    if ((err as CodedError).code !== "ENOENT") {
+      throw err;
+    }
+    return [];
+  }
+  const instances: DetachedInstance[] = [];
+
+  for (let i = 0; i < files.length; i++) {
+    const instanceName = files[i];
+    let shouldRemoveFile = false;
+
+    try {
+      // getDetachedInstanceByName() throws if the instance file is not found or
+      // cannot be parsed
+      const instance = await getDetachedInstanceByName(instanceName);
+
+      const matchingProcess = processes.find(p => p.pid === instance.pid);
+      if (!matchingProcess) {
+        console.warn(
+          `Process not found; ${instanceName} has been removed.  (PID: ${instance.pid}).`
+        );
+        shouldRemoveFile = true;
+      } else if (matchingProcess.cmd !== instance.cmd) {
+        // if the cmd does not match the instance, the process has been killed,
+        // and another application has taken the pid
+        console.warn(
+          `Process information doesn't match; ${instanceName} has been removed. (PID: ${instance.pid}).`
+        );
+        shouldRemoveFile = true;
+      } else {
+        instances.push(instance);
+      }
+    } catch (err) {
+      console.warn(
+        `Failed to load instance data; ${instanceName} has been removed.`
+      );
+      shouldRemoveFile = true;
+    }
+
+    if (shouldRemoveFile) removeDetachedInstanceFile(instanceName);
```

Comment:
Since `removeDetachedInstanceFile` is async, should this be `await`ed here? As is, if the deletion stalls for whatever reason, (/shrug), it will prevent the process from shutting down, and if it it throws it will _crash_ the process (though I know that `removeDetachedInstanceFile` wraps in a `try`/`catch`, so in this case it's impossible -- so throwing isn't really a concern)

Reply:
yes!